presto-parser/src/main/java/com/facebook/presto/sql/parser/ErrorHandler.java

```diff
@@ -0,0 +1,207 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.parser;
+
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.Parser;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+import org.antlr.v4.runtime.RuleContext;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.TokenStream;
+import org.antlr.v4.runtime.Vocabulary;
+import org.antlr.v4.runtime.atn.ATN;
+import org.antlr.v4.runtime.atn.ATNState;
+import org.antlr.v4.runtime.atn.NotSetTransition;
+import org.antlr.v4.runtime.atn.RuleStopState;
+import org.antlr.v4.runtime.atn.RuleTransition;
+import org.antlr.v4.runtime.atn.Transition;
+import org.antlr.v4.runtime.atn.WildcardTransition;
+import org.antlr.v4.runtime.misc.IntervalSet;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.antlr.v4.runtime.atn.ATNState.BLOCK_START;
+import static org.antlr.v4.runtime.atn.ATNState.RULE_START;
+
+class ErrorHandler
+        extends BaseErrorListener
+{
+    private final Map<Integer, String> specialRules;
+    private final Map<Integer, String> specialTokens;
+    private final Set<Integer> ignoredRules;
+
+    public ErrorHandler(Map<Integer, String> specialRules, Map<Integer, String> specialTokens, Set<Integer> ignoredRules)
+    {
+        this.specialRules = new HashMap<>(specialRules);
+        this.specialTokens = specialTokens;
+        this.ignoredRules = new HashSet<>(ignoredRules);
+    }
+
+    @Override
+    public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String message, RecognitionException e)
+    {
+        Parser parser = (Parser) recognizer;
+
+        ATN atn = parser.getATN();
+        ATNState currentState = atn.states.get(parser.getState());
+
+        Analyzer analyzer = new Analyzer(parser.getTokenStream(), atn, parser.getVocabulary(), specialRules, specialTokens, ignoredRules);
+
+        Map<Integer, Set<String>> candidates = new HashMap<>();
+        analyzer.process(candidates, currentState, parser.getCurrentToken().getTokenIndex(), parser.getContext());
+
+        // pick the candidate tokens associated largest token index processed (i.e., the path that consumed the most input)
+        String expected = candidates.entrySet().stream()
+                .max(Comparator.comparing(Map.Entry::getKey))
```

Comment:
should we pre-seed the candidates map with `0 -> something` to ensure we don't get an empty map?

Reply:
I don't think that could ever happen, unless there's a bug in the implementation. I'm adding some error handling to catch any unexpected errors.